R/iterate.R

```diff
@@ -9,6 +9,62 @@
 #' [iteration helper][iterate_with_offset] and use a
 #' [multi-response handler][resps_combine] to process the result.
 #'
+#' ## `next_req()`
+#'
+#' The key piece that makes `req_perform_iteratively()` work is the `next_req()`
+#' argument. For most common cases, you can use one of the canned helpers,
+#' like [iterate_with_offset()]. If, however, the API you're wrapping uses a
+#' different pagination system, you'll need to write your own. This section
+#' gives some advice.
+#'
+#' Generally, your function needs to inspect the response, extract some data
+#' from it, then use that to modify the previous request. For example, imagine
+#' that the response returns a cursor, which needs to be add to the body of
+#' the request. The simplest version of this function might look like this:
+#'
+#' ```R
+#' next_req <- function(resp, req) {
+#'   cursor <- resp_body_json(resp)$next_cursor
+#'   req %>% req_body_json_modify(cursor = cursor)
+#' }
+#' ```
+#'
+#' There's one problem here: if there are no more pages to return then
+#' `cursor` will be `NULL`, but `req_body_json_modify()` will still generate
+#' a meaningful request. So we need to handle this specifically by
+#' returning `NULL`:
+#'
+#' ```R
+#' next_req <- function(resp, req) {
+#'   cursor <- resp_body_json(resp)$next_cursor
+#'   if (is.null(cursor))
+#'     return(NULL)
+#'   req %>% req_body_json_modify(cursor = cursor)
+#' }
+#' ```
+#'
+#' A value of `NULL` lets `req_perform_iteratively()` know there are no pages
+#' remaining.
+#'
+#' There's one last feature you might want to add to your iterator: if you
+#' know the total number of pages, then it's nice to let
+#' `req_perform_iteratively()` know so it can adjust the progress bar.
+#' (This will only ever decrease the number of pages, not increase it.)
+#' You can signal the total number of pages by calling [signal_total_pages()],
+#' like this:
+#'
+#' ```R
+#' next_req <- function(resp, req) {
+#'   body <- resp_body_json(resp)
+#'   cursor <- body$next_cursor
+#'   if (is.null(cursor))
+#'     return(NULL)
+#'
+#'   signal_total_pages(body$pages)
```

Comment:
I don't think it's really that important to only signal this once. We could probably remove it from  `iterate_with_offset()` but I feel like it's worth being a bit more cautious in a helper, but it's generally hard to imagine that this is going to be expensive to compute.